cni/nnsgrpcclient/nnsgrpcclient_test.go

```diff
@@ -0,0 +1,117 @@
+package nnsgrpcclient
+
+import (
+	"context"
+	"fmt"
+	"github.com/Azure/azure-container-networking/test/nnsmockserver"
+	"os"
+	"strings"
+	"testing"
+	"time"
+)
+
+var mockserver *nnsmockserver.NnsMockServer
+
+func TestMain(m *testing.M) {
+	setup()
+	code := m.Run()
+	teardown()
+	os.Exit(code)
+}
+
+func setup() {
+	mockserver  = nnsmockserver.NewNnsMockServer()
+	go  mockserver.StartGrpcServer(nnsPort)
+	fmt.Printf("mock nns server started\n")
+}
+
+func teardown() {
+	mockserver.StopGrpcServer()
+	fmt.Printf("mock nns server stopped\n")
+}
+
+// CNI ADD to add container to network
+func TestAddContainerNetworking(t *testing.T) {
+
+	client := &NnsGrpcClient{}
+	if err := client.AddContainerNetworking(
+		 context.Background(),
+		"sf_8e9961f4-5b4f-4b3c-a9ae-c3294b0d9681",
+		"testnwspace"); err != nil {
+       t.Fatalf("TestAddContainerNetworking failed: %v", err)
+	}
+
+	fmt.Printf("TestAddContainerNetworking success\n")
+}
+
+// CNI DEL to delete container from network
+func TestDeleteContainerNetworking(t *testing.T) {
+
+	client := &NnsGrpcClient{}
+	if err := client.DeleteContainerNetworking(
+		context.Background(),
+		"sf_8e9961f4-5b4f-4b3c-a9ae-c3294b0d9681",
+		"testnwspace"); err != nil {
+		t.Fatalf("TestSetupContainer: %v", err)
+	}
+
+	fmt.Printf("TestDeleteContainerNetworking: success\n")
+}
+
+// CNI ADD to add container to network - failure case
+func TestAddContainerNetworkingFailure(t *testing.T) {
+
+	client := &NnsGrpcClient{}
+
+	var err error
+	if err = client.AddContainerNetworking(context.Background(), "testpod", "testnwspace"); err == nil {
+		t.Fatalf("TestAddContainerNetworkingFailure failed. Expected error but none returned\n")
+	}
+
+	if !strings.Contains(err.Error(), "Setup") {
+		t.Fatalf("TestAddContainerNetworkingFailure failed. Error should have contained Setup. %v \n", err)
+	}
+
+	fmt.Printf("TestAddContainerNetworkingFailure success\n")
+}
+
+// CNI DEL to add container to network - failure case
+func TestDeleteContainerNetworkingFailure(t *testing.T) {
+
+	client := &NnsGrpcClient{}
+
+	var err error
+	if err = client.DeleteContainerNetworking(context.Background(), "testpod", "testnwspace"); err == nil {
+		t.Fatalf("TestDeleteContainerNetworkingFailure failed. Expected error but none returned\n")
+	}
+
+	if !strings.Contains(err.Error(), "Teardown") {
+		t.Fatalf("TestDeleteContainerNetworkingFailure failed. Error should have contained Teardown. %v \n", err)
+	}
+
+	fmt.Printf("TestDeleteContainerNetworkingFailure success\n")
+}
+
+// CNI ADD to add container to network - grpc server was down temporarily and client reconnects well
+func TestAddContainerNetworkingGrpcServerDown(t *testing.T) {
+
+	// shutdown server to simulate connection error
+	teardown()
+
+	// bring up the server again after 25 seconds
+	go func() {
+		time.Sleep(25 * time.Second)
+		setup()
+	}()
```

Comment:
how does this test work? does the client retry indefinitely? if it doesn't, then since the mock server is shared in many tests, this can leave set up broken for future tests. If it does, 25 seconds seems very long for this test to run (and for the client to retry). let me know if im missing something

Reply:
client's connection timeout is 2 minutes, so it will retry for up 2 mins, so 25 seconds chosen here well within that range. If we run tests in parallel in this package, this could be a problem but in general we run it sequentially, so it shouldn't impact future tests.